Feature/multi rhs

maddyscientist opened the pull request:
This PR is focussed on framework evolution to get us ready for multi-RHS workflows
* Creation new `vector_ref` class which replaces the prior use of `std::vector<ColorSpinorField_ref>`
* Added multi-RHS interface to `Dirac` and `DiracMatrix` classes
* Move all regular Krylov solvers to use `std::vector<ColorSpinorField>` instead of `std::vector<ColorSpinorField*>`
* Significant cleanup of eigensolvers to use new multi-RHS framework
  * Block Lanczos eigen solver is now multi-RHS ready
* Created new class `FieldTmp` which provides a cache for temporary objects.  Deployed in the `Dirac` operators to negate the need to allocate and maintain explicit temporary vectors
* Add `max` and `max_deviation` reduce kernels, useful for correctness testing
* Added new test, `io_test`, for testing both gauge and color-spinor storage when QIO is enabled
* Cleanup of the blas functions: blas functions are now `const` correct, and doxygen is now added for all functions.
* Add deflation to BiCGStab
* Fix PCG solver which was broken when no preconditioner is present
* Fix various issues with ARPACK (lack of SVD support, correct e-vector extraction with non-Hermitian matrices)
* Various other misc fixes and cleanup

**Update**

Some additional fixes added during review:
* Clean up PCG and fix some corner cases
* Tweak `max_res_increase` default parameter to ensure more reliable convergence in ctest
* Add PCG and Schwarz preconditioned solvers to gtest list
* Fix compiler warnings when `QUDA_INVERFACE_NVTX=ON`
* Fix initialization error when using split grid
* Fix issue with BiCGStab

maddyscientist committed:
Add initial multi-RHS interface for Dirac and DiracMatrix classes

maddyscientist committed:
tunecache is not auto-dumped if the time since last tuning exceeds 120 seconds.  This allows us to clean up various calls to saveTuneCache()

maddyscientist committed:
Some initial cleanup of the eigen solver

maddyscientist committed:
Minor cleanup of multi_reduce.cu

maddyscientist committed:
Add support for non-sum reductions in blas reduce functors.  Add two new function max and max_deviation

maddyscientist committed:
Update is_enabled precision helper to use function arg instead of template - make it more useful

maddyscientist committed:
Add support to make_set to handle std::vector<T*> to std::vector<reference_wrapper> conversion

maddyscientist committed:
Some dilution test cleanup

maddyscientist committed:
Update VectorIO::load/save functions to internally use vector of ref_wrappers with a generic multi-rhs interface and a cleanup of the internals

maddyscientist committed:
math_helper should be included in blas_helper to ensure it is included

maddyscientist committed:
Fix compilation warning

maddyscientist committed:
Move some QUDA feature enablement to quda_define.h.in instead of requiring compiler flags

maddyscientist committed:
Added new test file io_test: this does unit testing on i/o functionlity for both gauge and color-spinor fields

maddyscientist committed:
deflation.cpp should use VectorIO class for load/saving vectors

maddyscientist committed:
Simplify eigensolver.cpp

maddyscientist committed:
Replace the blas ax_ functor with axy_ which allows for reading from and writing to distinct fields.  blas::ax is now just a wrapper around this, and introduce blas::axy which exposes this functionality

maddyscientist committed:
Improve error verbosity in multi-BLAS _U/_L functions

maddyscientist committed:
Add ColroSpinorField_cref alias for const fields

maddyscientist committed:
Update multi-blas interface function to accept vectors of ColorSpinorField_cref instead of ColorSpinorField_ref where the vectors are read only

maddyscientist committed:
Mark ColorSpinorField/LatticeField copy and move constructors to be noexcept - this allows for move assignment instead of copy assignment when std::vector::resize is called

maddyscientist committed:
Fix issue with 5bf5c3f66f48ac9bc36edae19de915b386548586, where for functions that who require a vector of crefs, we would endup self recursing if passing a vector of refs: added explicit make_cset helper, to create explicit cref sets to avoid this

maddyscientist committed:
Add custom resize functions for std::vector<ColorSpinorField> instances that allow us to avoid the copying that would occur if we used std::vector::resize

maddyscientist committed:
Use custom resize function in blas_test and dslash_test: this significantly boosts the performance of the blas_test, removing a lot of copies

maddyscientist committed:
Rewrite of eigensolvers to utilize vector of fields instead of vector of pointers.  As part of this change, the multi-blas function interfaces have been overhauled to utilized a new specialized type vector_ref, to provide a universal interface

maddyscientist committed:
Merge branch 'hotfix/arpack' of github.com:lattice/quda into feature/multi-rhs

cpviolator committed:
Add SVD computation to ARPACK path

cpviolator committed:
Merge branch 'feature/multi-rhs' of https://github.com/lattice/quda into feature/multi-rhs

maddyscientist committed:
Fix resize bug

cpviolator committed:
Add function to allow for ARPACK sorting, use 'A' param to return Ritz vectors in ARPACK, not Schur vectors

cpviolator committed:
Merge branch 'feature/multi-rhs' of https://github.com/lattice/quda into feature/multi-rhs

maddyscientist committed:
Promote block axpy functions to be templated on coefficient type: this allows their use in generic code that deals with real or complex-valued arithmetic

maddyscientist committed:
The member functions rotateVecs and blockRotate are now generic, and work with real or complex arithmetic

cpviolator committed:
Minor clean up of ARPACK eigen/singular value computation.

cpviolator committed:
Minor clean up of ARPACK eigen/singular value computation.

maddyscientist committed:
Add init member to LatticeFieldParam, which is used to check if the parameter structure is trivially constructed or not.  Use this to simplify the ColorSpinorField::create_alias interface

maddyscientist committed:
Generalizer pair constructor for reference_wrapper to a variadic constructor

maddyscientist committed:
Remove all use of std::vector<ColorSpinorField*> from regular solvers

maddyscientist committed:
Merge branch 'feature/multi-rhs' of github.com:lattice/quda into feature/multi-rhs

maddyscientist committed:
Delete accidental inclusion

maddyscientist committed:
Minor fix for max_deviation functor

maddyscientist committed:
Added FieldTmp class for caching temporaries fields.  Removed all explicit management of temporaries relating to the Dirac operators (significant cleanup)

maddyscientist committed:
Add missing header from last commit

maddyscientist committed:
Remove ColorSpinorField_ref type: replaced all usage using vector_ref

maddyscientist committed:
Update multi-RHS Dirac interface to use vector_ref

maddyscientist committed:
Cleanup for blas functions: added missing doxygen, all blas function arguments are now const where they should be, axpyCGNorm now returns a double2 instead of Complex

maddyscientist committed:
Remove now unnecessary const_cast usage when calling blas functions - enabled by last commit

maddyscientist committed:
Initial multi-RHS awareness added to block Lanczos

cpviolator committed:
Add deflation to bicgstab

cpviolator committed:
Merge branch 'feature/multi-rhs' of https://github.com/lattice/quda into feature/multi-rhs

cpviolator committed:
Remove 'mat' argument from eigesolver member functions, use 'mat' reference instead

maddyscientist committed:
Make blas::zero function multi-RHS aware.  Fix bug in EigenSolver::rotateVecs function

maddyscientist committed:
Fix PCG when no preconditioner is used

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/multi-rhs

maddyscientist committed:
Fix issue with field cache: needs to be nuked if the comm partitioning is changed

maddyscientist committed:
Remove last remnants of Dirac::tmp1 / tmp2

maddyscientist committed:
Fix for VectorIO interface, and add WAR for multigrid's use of VectorIO (will be fixed when MG uses vector of fields, instead of vector of pointers)

maddyscientist committed:
Add missing string header inclusion

maddyscientist committed:
Fix Jenkins error with BiCGStab

maddyscientist committed:
Add PCG to the invert_test list

maddyscientist committed:
Upate multi-RHS interface functions to use const vector_ref lvalue instead of vector_ref rvalue.  Add array accessor for vector_ref, and apply the resulting simplifications.

maddyscientist committed:
Created cvector_ref alias for const vector_ref

maddyscientist committed:
Fix some long-standing issues in PCG, and apply some cleanup

maddyscientist committed:
Tweak default parameter max_res_increase to ensure convergence for all ctest solver test

maddyscientist committed:
Fix bug when running io_test with multiple processes

maddyscientist committed:
Fix bug with dslash_ctest: test_split_grid is not yet initialized when checking if kernel will skip (fixes occasional hangs when running on multiple processes)

mathiaswagner commented on include/field_cache.h:
```diff
@@ -0,0 +1,42 @@
+#include <map>
+#include <stack>
+
+namespace quda {
```
THis file needs some doxygen

maddyscientist replied:
[b37ed336](https://github.com/lattice/quda/commit/b37ed336c69a82e851d2cfccd3ca5472c750f718)

mathiaswagner commented on include/quda_define.h.in:
```diff
@@ -29,6 +29,123 @@
 #undef QUDA_HETEROGENEOUS_ATOMIC
 #endif
 
+#cmakedefine QUDA_DIRAC_NDEG_TWISTED_CLOVER
```
As discussed previously I am not a big fan of a monolithic define file for recompile/ ccache efficiency. Happy to have that for apps that use QUDA but we should aim for something more fine grained. But for another PR.

maddyscientist replied:
Yep, understood.  This change doesn't change the needle in any regard, it moves the prior compiler flags to a preconfigured flag.

I made this change to fix bugs in test codes: the enablement flags were not passed to the tests, which meant that some functions would evaluate differently internally and externally of QUDA.

mathiaswagner replied:
Makes sense. We need to put some work in to really improve here.

mathiaswagner commented on include/tunable_reduction.h:
```diff
@@ -71,7 +71,7 @@ namespace quda
     void launch_device(T &result, const TuneParam &tp, const qudaStream_t &stream, Arg &arg)
     {
-      if (tp.block.x * tp.block.y < device::warp_size())
+      if (tp.block.x * tp.block.y < (unsigned)device::warp_size())
```
use C++ cast

maddyscientist replied:
[b37ed336](https://github.com/lattice/quda/commit/b37ed336c69a82e851d2cfccd3ca5472c750f718)

mathiaswagner commented on include/tunable_reduction.h:
```diff
@@ -226,7 +226,7 @@ namespace quda
     void launch_device(std::vector<T> &result, const TuneParam &tp, const qudaStream_t &stream, Arg &arg)
     {
-      if (tp.block.x * tp.block.y < device::warp_size())
+      if (tp.block.x * tp.block.y < (unsigned)device::warp_size())
```
C++ cast

maddyscientist replied:
[b37ed336](https://github.com/lattice/quda/commit/b37ed336c69a82e851d2cfccd3ca5472c750f718)

mathiaswagner commented on include/dirac_quda.h:
```diff
@@ -2526,48 +2429,21 @@ namespace quda {
-        dirac->tmp1 = NULL;
-        reset1 = false;
+      for (auto i = 0u; i < in.size(); i++) {
+        if (shift != 0.0) blas::axpy(shift, in[i], out[i]);
```
just wondering if we can move the if(shift != 0) out of the for loop

maddyscientist replied:
This is just temporary code.  Long-term plan here is to update the `single-RHS` BLAS functions to be batched functions (as opposed to block BLAS).  So we'd be able to write `blas::axpy(shift, in, out)` and it would do the correct thing.  Though doing so may require we namespace the block BLAS to avoid collision between batched and block?

mathiaswagner commented on include/blas_quda.h:
```diff
@@ -26,126 +26,458 @@ namespace quda {
-    void zero(ColorSpinorField &a);
+    inline void zero(cvector_ref<ColorSpinorField> &x)
+    {
+      for (auto &xi : x) xi.get().zero();
```
just wondering if for cases like that we could come up with a way to do something to get rid of the .get(). But maybe there just is no easy/nice/clean way around this here.

maddyscientist replied:
If we use the array accessor, we can avoid the `get()` since I overloaded it return the underlying reference.

[b37ed336](https://github.com/lattice/quda/commit/b37ed336c69a82e851d2cfccd3ca5472c750f718)

mathiaswagner committed:
fix a few typos

mathiaswagner commented on lib/dirac_clover.cpp:
```diff
@@ -59,14 +59,10 @@ namespace quda {
   void DiracClover::MdagM(ColorSpinorField &out, const ColorSpinorField &in) const
   {
     checkFullSpinor(out, in);
+    auto tmp = getTmp(in);
```
looking at this now I don't think getTmp with this usage is a good name. It get's some TMP that is stored in an auto variable. How about making it a bit more obvious that we are getting a tmp ColorSpinorField?

maddyscientist replied:
The object returned here isn't a `ColorSpinorField` and the `auto` was intentional here to avoid having to explicitly state what it was (`FieldTmp<ColorSpinorField>`).

What are your thoughts here?

mathiaswagner replied:
can we rename `getTmp` to `getTmpField` or something like that? Thinking out lout. Just aiming for `the code is the documentation` but otherwise it is not the end of the world as this is not a code touched a lot.

maddyscientist replied:
I can do that 👍

maddyscientist replied:
[b37ed336](https://github.com/lattice/quda/commit/b37ed336c69a82e851d2cfccd3ca5472c750f718)

mathiaswagner commented on lib/color_spinor_util.cu:
```diff
@@ -424,4 +424,29 @@ namespace quda {
     }
   }
 
+  void resize(std::vector<ColorSpinorField> &v, size_t new_size, const ColorSpinorParam &param)
```
not related to this change but wondering if color_cpinor_util could become a cpp file

maddyscientist replied:
That is my plan.  It's not possible at the moment because a variety of the headers included here expect a device aware compiler.  E.g., `math_helper.cuh` doesn't work if included in gcc.  I intend to fix this....

mathiaswagner commented on lib/multigrid.cpp:
```diff
@@ -1832,12 +1807,9 @@ namespace quda
     for (int i = 0; i < (int)param.B.size(); i++) {
       param.B[i] = new ColorSpinorField(bParam);
-      *param.B[i] = *B_evecs[i];
+      *param.B[i] = B_evecs[i]; // FIXME can std::move this
```
fix now or later?

maddyscientist replied:
Leave it for later.  The multigrid.cpp handling of vector sets needs to be rewritten completely, and much smarter things can be done here.  I intentionally tried to minimize touching multigrid.cpp to minimize conflicts with #1283.

Moreover, that specific code needs to made more robust (e.g., what happens if `n_conv` isn't at least equal to `B.size()`?).  

My plan is to do a specific multigrid-focused cleanup PR that will also add proper regression tests.

mathiaswagner commented:
There are still a couple of saveTuneCache in the interface. I guess they can all be deleted?

hummingtree commented on include/eigensolve_quda.h:
```diff
@@ -237,25 +222,11 @@ namespace quda
-    void blockRotateComplex(std::vector<ColorSpinorField *> &kSpace, Complex *array, int rank, const range &i,
-                            const range &j, blockType b_type, int offset);
+    template <typename T, typename matrix_t>
+    void blockRotate(std::vector<ColorSpinorField> &kSpace, matrix_t &array, int rank,
+                     const range &i, const range &j, blockType b_type, int offset);
```
Is there a general guidance on when to use `std::vector<ColorSpinorField> &` in the intefaces and when to use `cvector_ref<ColorSpinorField> &` instead? The former obviously owns the fields, but I am asking more from an interface point of view.

maddyscientist replied:
This will be a long-term evolution, and beyond the scope of this PR....

I'm thinking that we'll want to favour the use of `cvector_ref<ColorSpinorField>` since it will enable multi-RHS functionality with little effort.  The next step here will be in my follow up PR that adds actual multi-RHS Dirac kernels, and I will continue to push in this direction then.

For example, I think that we may want to upgrade all BLAS functions to accept `cvector_ref<>` arguments.  This would better enable us to have multi-RHS solvers everywhere.

maddyscientist replied:
And I realize I misread the question. 😄 (was thinking it was `cvector_ref` vs `ColorSpinorField&`)

I think in general using `cvector_ref` should be preferred for lower-level functions, e.g., below the level of solvers.  At the solver level, I could imagine we'll want both, and above the solvers I imagine using `std::vector` is the way to go?  🤷‍♀️ 

Again, this can be something that can be long-term evolution, as we add more multi-RHS functionality we will work out the answer.

maddyscientist commented:
> There are still a couple of saveTuneCache in the interface. I guess they can all be deleted?

I left those intentionally, but happy to discuss their removal.  Those were left to ensure that when the solver finishes, the tunecache is dumped which I naively thought was desirable.  Given it would be dumped anyway when `endQuda` is called, I'm happy to delete for code cleanup.

Thoughts?

maddyscientist committed:
Address Mathias' review comments

maddyscientist committed:
Add Schwarz solvers to the invert_test gtest list.  Change default CA basis used in preconditioners to Chebyshev

mathiaswagner commented on include/eigensolve_quda.h:
```diff
@@ -136,34 +133,32 @@ namespace quda
        @param[in] in Input spinor
     */
-    void matVec(const DiracMatrix &mat, ColorSpinorField &out, const ColorSpinorField &in);
+    //void chebyOp(const DiracMatrix &mat, ColorSpinorField &out, const ColorSpinorField &in);
```
can this line be deleted?

maddyscientist replied:
6df70599df5a0faaec4a5e894c113898698f2dda

maddyscientist committed:
Fix bug in invert_test, introduced when adding Schwarz preconditioner support

maddyscientist committed:
Delete dead code

maddyscientist committed:
Merge branch 'feature/multi-rhs' of github.com:lattice/quda into feature/multi-rhs

mathiaswagner commented on tests/invert_test_gtest.hpp:
```diff
@@ -126,15 +136,21 @@ std::string gettestname(::testing::TestParamInfo<test_t> param)
-  if (::testing::get<5>(param.param) > 1)
-    name += std::string("_solution_accumulator_pipeline_") + std::to_string(::testing::get<5>(param.param));
+    name += std::string("_shift") + std::to_string(::testing::get<4>(param.param));
+  if (::testing::get<5>(param.param) > 1)// && support_solution_accumulator_pipeline(::testing::get<0>(param.param)))
```
what about the commented part here?

maddyscientist replied:
Erroneous inclusion.  Fixed with [cdc47f9](https://github.com/lattice/quda/pull/1318/commits/cdc47f9163faa35e24b29d43bfe32de6da981c4c)

mathiaswagner committed:
clean up headers, fix typo

maddyscientist committed:
Remove commented code

maddyscientist committed:
Merge branch 'feature/multi-rhs' of github.com:lattice/quda into feature/multi-rhs

hummingtree commented on lib/madwf_ml.cpp:
```diff
@@ -222,10 +219,8 @@ namespace quda
-        Complex dot[9];
-        blas::cDotProduct(dot, lhs, rhs);
+        std::vector<Complex> dot(9);
+        blas::cDotProduct(dot, {chi, theta, lambda}, {chi, theta, lambda});
```
I love that fact that we can do this directly now. :)

mathiaswagner reviewed (approved):
Visual review looks good, but I have not done any testing beyond a simple build.

hummingtree commented on lib/blas_quda.cu:
```diff
@@ -26,19 +26,19 @@ namespace quda {
-        z(z),
-        w(w),
-        v(v)
+        x(const_cast<ColorSpinorField&>(x)),
+        y(const_cast<ColorSpinorField&>(y)),
+        z(const_cast<ColorSpinorField&>(z)),
+        w(const_cast<ColorSpinorField&>(w)),
+        v(const_cast<ColorSpinorField&>(v))
```
This is out of the scope of this PR, but we need to have these `const_cast`'s to have only one definition of `Blas` and `BlasArg` right? Is there a template trick that would get rid of them?

maddyscientist replied:
It's definitely the plan to remove these `const_cast`s.  The immediate issue that I haven't solved yet is that we don't know the type of the input `x`, `y`, etc. fields since they are template arguments to the constructor so when declaring the types in the class there's no way to know if they should be const or not.

hummingtree replied:
Maybe some sort of variadic templated `Blas` would do? I will think more about this.

maddyscientist replied:
👍

hummingtree commented on lib/communicator_stack.cpp:
```diff
@@ -167,13 +167,17 @@ namespace quda
   }
 
-  void comm_free(MsgHandle *&mh) { get_current_communicator().comm_free(mh); }
+#define CHECK_MH(mh) { if (mh == nullptr) errorQuda("null message handle"); }
```
Can we make this a function instead of a macro?

maddyscientist replied:
I had this as a macro so that the correct line and function location will be reported.  I think we should leave this as a macro until we can jump to C++20 (`std::source_location`)

hummingtree commented on lib/dirac.cpp:
```diff
@@ -312,8 +289,6 @@ namespace quda {
   void Dirac::prefetch(QudaFieldLocation mem_space, qudaStream_t stream) const
   {
     if (gauge) gauge->prefetch(mem_space, stream);
-    if (tmp1) tmp1->prefetch(mem_space, stream);
-    if (tmp2) tmp2->prefetch(mem_space, stream);
```
So now with the dynamic temporary fields how is the prefetch done?

maddyscientist replied:
That's a good question.  I could add a prefetch (if enabled) whenever the temporary is requested.

maddyscientist replied:
Having thought about this a bit more, I'm going to leave it as is.  We haven't really found the need for prefetching in production yet.  If we need to consider prefetching in the future here for temporaries, we can come back to that in the future.

weinbe2 replied:
Staggered MG is a reasonably production reason, but spinor fields generally aren't beefy enough to matter --- this comment is more "for the record/awareness" than a "we need to discuss this now" type deal.

maddyscientist replied:
Acknowledge and agreed

hummingtree commented on lib/dirac_domain_wall_4d.cpp:
```diff
@@ -158,42 +155,38 @@ namespace quda {
-    deleteTmp(&tmp2, reset);
+    auto tmp = getFieldTmp(in);
+    M(tmp, in);
+    Mdag(out, tmp);
```
This is great in the sense that we do not need to distinguish between `tmp1` and `tmp2` anymore. :)

hummingtree commented on lib/multi_blas_quda.cu:
```diff
@@ -557,7 +571,7 @@ namespace quda {
     {
       std::vector<double> a_(x.size() * y.size());
       memcpy(a_.data(), a, x.size() * y.size() * sizeof(double));
-      std::vector<std::reference_wrapper<ColorSpinorField>> x_;
+      std::vector<std::reference_wrapper<const ColorSpinorField>> x_;
```
Here and later in this file, why not use `vector_ref`?

maddyscientist replied:
[b607f03](https://github.com/lattice/quda/pull/1318/commits/b607f03e157ea8c08aae26a419c4c56ad5d6f40a)

hummingtree reviewed (approved):
The changes look great. I will do some actual builds.

maddyscientist committed:
Fix compiler warnings when QUDA_INVERFACE_NVTX=ON

maddyscientist committed:
Remove some legacy std::vector<std::reference_wrapper<ColorSpinorField>>

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/multi-rhs

weinbe2 reviewed (approved):
Visual only, but there's a lot of good cleanup here, and I appreciate the additional unit testing. Since others have been hands-on with this PR, I'm going to give it the approval based on my visual review. Awesome work!

maddyscientist commented:
Merging this now.  I've verified manually:
* split grid solvers are working as expected
* multigrid file i/o

I've updated the description with a few late additions to this PR that were added during review.
